File: JarProtoExtractor/src/main/java/cyt/jarprotoextractor/ProtoGenerator.java
package cyt.jarprotoextractor;

import java.lang.reflect.Method;
import java.util.List;
import java.util.Set;
import java.util.HashSet;

public class ProtoGenerator {
    private static final String MAP_ENTRY_SUFFIX = "Entry";

    private Class<?> fileDescriptorClass;
    private Class<?> descriptorClass;
    private Class<?> fieldDescriptorClass;
    private Class<?> enumDescriptorClass;
    private Class<?> enumValueDescriptorClass;
    private Class<?> oneofDescriptorClass;
    private Class<?> javaTypeClass;

    public boolean loadProtobufRuntime(ClassLoader commonClassLoader) {
        try {
            fileDescriptorClass = commonClassLoader.loadClass("com.google.protobuf.Descriptors$FileDescriptor");
            descriptorClass = commonClassLoader.loadClass("com.google.protobuf.Descriptors$Descriptor");
            fieldDescriptorClass = commonClassLoader.loadClass("com.google.protobuf.Descriptors$FieldDescriptor");
            enumDescriptorClass = commonClassLoader.loadClass("com.google.protobuf.Descriptors$EnumDescriptor");
            enumValueDescriptorClass = commonClassLoader.loadClass("com.google.protobuf.Descriptors$EnumValueDescriptor");
            oneofDescriptorClass = commonClassLoader.loadClass("com.google.protobuf.Descriptors$OneofDescriptor");
            javaTypeClass = commonClassLoader.loadClass("com.google.protobuf.Descriptors$FieldDescriptor$JavaType");

            return true;
        } catch (Exception e) {
            System.err.println("Failed to load Protobuf runtime: " + e.getMessage());
            e.printStackTrace();
            return false;
        }
    }

    public String generateProtoContent(List<Class<?>> messageClasses) throws Exception {
        StringBuilder protoContent = new StringBuilder("syntax = \"proto3\";\n\n");
        Set<Object> processedEnums = new HashSet<>();
        Set<Object> processedMessages = new HashSet<>();

        for (Class<?> clazz : messageClasses) {
            Object fileDescriptor = getDescriptor(clazz);
            if (fileDescriptor != null) {
                protoContent.append("// ").append(clazz).append("\n");
                processFileDescriptorEnums(fileDescriptor, protoContent, processedEnums);

                Method getMessageTypes = fileDescriptorClass.getMethod("getMessageTypes");
                @SuppressWarnings("unchecked")
                List<Object> messageDescriptors = (List<Object>) getMessageTypes.invoke(fileDescriptor);

                for (Object messageDescriptor : messageDescriptors) {
                    if (processedMessages.add(messageDescriptor)) {
                        generateMessage(messageDescriptor, protoContent, 0, true, processedEnums);
                        protoContent.append("\n");
                    }
                }
            }
        }
        return protoContent.toString();
    }

    private void processFileDescriptorEnums(Object descriptor,
                                            StringBuilder sb,
                                            Set<Object> processed) throws Exception {
        Method getEnumTypes = fileDescriptorClass.getMethod("getEnumTypes");
        @SuppressWarnings("unchecked")
        List<Object> enumDescs = (List<Object>) getEnumTypes.invoke(descriptor);

        for (Object enumDesc : enumDescs) {
            if (processed.add(enumDesc)) {
                generateEnum(enumDesc, sb, 0);
            }
        }
    }

    private void generateEnum(Object enumDesc,
                              StringBuilder sb,
                              int indentLevel) throws Exception {
        Method getName = enumDescriptorClass.getMethod("getName");
        String name = (String) getName.invoke(enumDesc);

        String indent = "    ".repeat(indentLevel);
        sb.append(indent).append("enum ").append(name).append(" {\n");

        Method getValues = enumDescriptorClass.getMethod("getValues");
        @SuppressWarnings("unchecked")
        List<Object> values = (List<Object>) getValues.invoke(enumDesc);

        for (Object value : values) {
            Method getValueName = enumValueDescriptorClass.getMethod("getName");
            String valueName = (String) getValueName.invoke(value);

            Method getValueNumber = enumValueDescriptorClass.getMethod("getNumber");
            int valueNumber = (int) getValueNumber.invoke(value);

            sb.append(indent).append("    ")
                    .append(valueName)
                    .append(" = ")
                    .append(valueNumber)
                    .append(";\n");
        }
        sb.append(indent).append("}\n\n");
    }

    private Object getDescriptor(Class<?> messageClass) throws Exception {
        try {
            Method getDescriptor = messageClass.getMethod("getDescriptor");
            return getDescriptor.invoke(null);
        } catch (NoSuchMethodException e) {
            return null;
        }
    }

    private void generateMessage(Object messageDescriptor,
                                 StringBuilder sb,
                                 int indentLevel,
                                 boolean isTopLevel,
                                 Set<Object> processedEnums) throws Exception {
        Method getName = descriptorClass.getMethod("getName");
        String name = (String) getName.invoke(messageDescriptor);

        String indent = "    ".repeat(indentLevel);
        sb.append(indent).append("message ").append(name).append(" {\n");

        Method getEnumTypes = descriptorClass.getMethod("getEnumTypes");
        @SuppressWarnings("unchecked")
        List<Object> enumDescs = (List<Object>) getEnumTypes.invoke(messageDescriptor);

        for (Object enumDesc : enumDescs) {
            if (processedEnums.add(enumDesc)) {
                generateEnum(enumDesc, sb, indentLevel + 1);
            }
        }

        Method getOneofs = descriptorClass.getMethod("getOneofs");
        @SuppressWarnings("unchecked")
        List<Object> oneofs = (List<Object>) getOneofs.invoke(messageDescriptor);

        for (Object oneof : oneofs) {
            Method getOneofName = oneofDescriptorClass.getMethod("getName");
            String oneofName = (String) getOneofName.invoke(oneof);

            sb.append(indent).append("    oneof ").append(oneofName).append(" {\n");

            Method getFields = oneofDescriptorClass.getMethod("getFields");
            @SuppressWarnings("unchecked")
            List<Object> fields = (List<Object>) getFields.invoke(oneof);

            for (Object field : fields) {
                appendField(field, sb, indentLevel + 2);
            }
            sb.append(indent).append("    }\n");
        }

        Method getFields = descriptorClass.getMethod("getFields");
        @SuppressWarnings("unchecked")
        List<Object> fields = (List<Object>) getFields.invoke(messageDescriptor);

        for (Object field : fields) {
            Method getContainingOneof = fieldDescriptorClass.getMethod("getContainingOneof");
            Object containingOneof = getContainingOneof.invoke(field);
            if (containingOneof != null) continue;

            appendField(field, sb, indentLevel + 1);
        }

        Method getNestedTypes = descriptorClass.getMethod("getNestedTypes");
        @SuppressWarnings("unchecked")
        List<Object> nestedTypes = (List<Object>) getNestedTypes.invoke(messageDescriptor);

        for (Object nested : nestedTypes) {
            String nestedName = (String) getName.invoke(nested);
            if (!nestedName.endsWith(MAP_ENTRY_SUFFIX)) {
                generateMessage(nested, sb, indentLevel + 1, false, processedEnums);
                sb.append("\n");
            }
        }

        sb.append(indent).append("}");
        if (isTopLevel) sb.append("\n");
    }

    private void appendField(Object field,
                             StringBuilder sb,
                             int indentLevel) throws Exception {
        String indent = "    ".repeat(indentLevel);

        Method isRepeated = fieldDescriptorClass.getMethod("isRepeated");
        boolean repeated = (boolean) isRepeated.invoke(field);

        Method isMapField = fieldDescriptorClass.getMethod("isMapField");
        boolean mapField = (boolean) isMapField.invoke(field);

        String modifier = "";
        if (!mapField && repeated) {
            modifier = "repeated ";
        }

        Method getName = fieldDescriptorClass.getMethod("getName");
        String name = (String) getName.invoke(field);

        Method getNumber = fieldDescriptorClass.getMethod("getNumber");
        int number = (int) getNumber.invoke(field);

        String typeName = getTypeName(field);

        if (mapField) {
            Method getMessageType = fieldDescriptorClass.getMethod("getMessageType");
            Object entryDesc = getMessageType.invoke(field);

            Method getFieldsMethod = descriptorClass.getMethod("getFields");
            @SuppressWarnings("unchecked")
            List<Object> entryFields = (List<Object>) getFieldsMethod.invoke(entryDesc);

            Object keyField = null;
            Object valueField = null;
            for (Object entryField : entryFields) {
                String fieldName = (String) getName.invoke(entryField);
                if ("key".equals(fieldName)) {
                    keyField = entryField;
                } else if ("value".equals(fieldName)) {
                    valueField = entryField;
                }
            }

            if (keyField != null && valueField != null) {
                String keyType = getTypeName(keyField);
                String valueType = getTypeName(valueField);
                typeName = String.format("map<%s, %s>", keyType, valueType);
            } else {
                typeName = "map<unknown, unknown>";
            }
        }

        String line = String.format("%s%s%s %s = %d;",
                indent,
                modifier,
                typeName,
                name,
                number);
        sb.append(line).append("\n");
    }

    private String getTypeName(Object field) throws Exception {
        Method getType = fieldDescriptorClass.getMethod("getType");
        Object fieldType = getType.invoke(field);

        Method nameMethod = fieldType.getClass().getMethod("name");
        String typeName = ((String) nameMethod.invoke(fieldType)).toUpperCase();

        Method getJavaType = fieldDescriptorClass.getMethod("getJavaType");
        Object javaType = getJavaType.invoke(field);

        Method javaTypeName = javaTypeClass.getMethod("name");
        String javaTypeStr = (String) javaTypeName.invoke(javaType);

        switch (javaTypeStr) {
            case "ENUM":
                Method getEnumType = fieldDescriptorClass.getMethod("getEnumType");
                Object enumType = getEnumType.invoke(field);
                Method getEnumTypeName = enumDescriptorClass.getMethod("getName");
                return (String) getEnumTypeName.invoke(enumType);
            case "MESSAGE":
                Method getMessageType = fieldDescriptorClass.getMethod("getMessageType");
                Object messageType = getMessageType.invoke(field);
                Method getMessageTypeName = descriptorClass.getMethod("getName");
                return (String) getMessageTypeName.invoke(messageType);
        }

        switch (typeName) {
            case "DOUBLE": return "double";
            case "FLOAT": return "float";
            case "INT64": return "int64";
            case "UINT64": return "uint64";
            case "INT32": return "int32";
            case "FIXED64": return "fixed64";
            case "FIXED32": return "fixed32";
            case "STRING": return "string";
            case "BYTES": return "bytes";
            case "UINT32": return "uint32";
            case "SFIXED32": return "sfixed32";
            case "SFIXED64": return "sfixed64";
            case "SINT32": return "sint32";
            case "SINT64": return "sint64";
            case "BOOL": return "bool";
            default: return typeName.toLowerCase();
        }
    }
}

File: proto_generator.py
from typing import Tuple, Optional
import google.protobuf.descriptor_pb2 as descriptor_pb2

def generate_proto_content(file_descriptor):
    lines = []

    syntax = file_descriptor.syntax if file_descriptor.syntax else "proto3"
    lines.append(f'syntax = "{syntax}";\n')

    if file_descriptor.package:
        lines.append(f"package {file_descriptor.package};\n")

    for dependency in file_descriptor.dependency:
        lines.append(f'import "{dependency}";')

    if file_descriptor.dependency:
        lines.append("")

    for enum in file_descriptor.enum_type:
        generate_enum(enum, lines, 0)

    for message in file_descriptor.message_type:
        generate_message(message, lines, 0)

    return "\n".join(lines)

def generate_enum(enum_desc, lines, indent_level):
    indent = "    " * indent_level
    lines.append(f"{indent}enum {get_simple_type_name(enum_desc.name)} {{")
    for value in enum_desc.value:
        lines.append(f"{indent}    {value.name} = {value.number};")
    lines.append(f"{indent}}}")
    lines.append("")

def generate_message(message_desc, lines, indent_level):
    indent = "    " * indent_level
    lines.append(f"{indent}message {get_simple_type_name(message_desc.name)} {{")

    generate_map_fields(message_desc, lines, indent_level)

    for field in message_desc.field:
        if field.HasField("oneof_index"):
            continue

        if is_map_field(field, message_desc):
            continue

        field_label = get_field_label(field)
        field_type = get_field_type(field)
        lines.append(f"{indent}    {field_label}{field_type} {field.name} = {field.number};")

    generate_oneof_fields(message_desc, lines, indent_level)

    generate_nested_types(message_desc, lines, indent_level)

    lines.append(f"{indent}}}")
    lines.append("")

def generate_map_fields(message_desc, lines, indent_level):
    indent = "    " * indent_level
    for nested in message_desc.nested_type:
        if nested.options.map_entry:
            map_field = find_map_field(nested, message_desc)
            if map_field:
                key_field, value_field = get_map_entry_fields(nested)
                key_type = get_field_type(key_field)
                value_type = get_field_type(value_field)
                lines.append(f"{indent}    map<{key_type}, {value_type}> {map_field.name} = {map_field.number};")

def is_map_field(field, message_desc):
    for nested in message_desc.nested_type:
        if nested.options.map_entry:
            if field.type_name.endswith(nested.name):
                return True
    return False

def generate_oneof_fields(message_desc, lines, indent_level):
    if not message_desc.oneof_decl:
        return

    indent = "    " * indent_level
    oneof_groups = group_oneof_fields(message_desc)

    for index, fields in oneof_groups.items():
        if index >= len(message_desc.oneof_decl):
            continue

        oneof_name = message_desc.oneof_decl[index].name
        lines.append(f"{indent}    oneof {oneof_name} {{")
        for field in fields:
            field_type = get_field_type(field)
            lines.append(f"{indent}        {field_type} {field.name} = {field.number};")
        lines.append(f"{indent}    }}")

def generate_nested_types(message_desc, lines, indent_level):
    for enum in message_desc.enum_type:
        generate_enum(enum, lines, indent_level + 1)
    for nested in message_desc.nested_type:
        if not nested.options.map_entry:
            generate_message(nested, lines, indent_level + 1)

def find_map_field(map_entry, parent_desc):
    for field in parent_desc.field:
        if field.type_name.endswith(map_entry.name):
            return field
    return None

def get_map_entry_fields(map_entry):
    return map_entry.field[0], map_entry.field[1]

def group_oneof_fields(message_desc):
    oneof_groups = {}
    for field in message_desc.field:
        if field.HasField("oneof_index"):
            index = field.oneof_index
            oneof_groups.setdefault(index, []).append(field)
    return oneof_groups

def get_field_label(field):
    if field.label == field.LABEL_REPEATED:
        return "repeated "
    if field.proto3_optional:
        return "optional "
    return ""

def get_field_type(field):
    if field.HasField("type_name"):
        return get_simple_type_name(field.type_name)

    type_mapping = {
        field.TYPE_DOUBLE: "double",
        field.TYPE_FLOAT: "float",
        field.TYPE_INT64: "int64",
        field.TYPE_UINT64: "uint64",
        field.TYPE_INT32: "int32",
        field.TYPE_FIXED64: "fixed64",
        field.TYPE_FIXED32: "fixed32",
        field.TYPE_BOOL: "bool",
        field.TYPE_STRING: "string",
        field.TYPE_BYTES: "bytes",
        field.TYPE_UINT32: "uint32",
        field.TYPE_SFIXED32: "sfixed32",
        field.TYPE_SFIXED64: "sfixed64",
        field.TYPE_SINT32: "sint32",
        field.TYPE_SINT64: "sint64",
    }
    return type_mapping.get(field.type, "unknown")

def get_simple_type_name(full_name):
    if full_name.startswith("."):
        full_name = full_name[1:]
    return full_name.split(".")[-1]

def generate_proto_from_bytes(descriptor_bytes: bytes) -> Tuple[str, Optional[str]]:

    file_descriptor = descriptor_pb2.FileDescriptorProto()
    file_descriptor.ParseFromString(descriptor_bytes)

    proto_content = generate_proto_content(file_descriptor)

    # name
    proto_filename = file_descriptor.name
    if proto_filename:
        proto_filename = proto_filename
    else:
        proto_filename = None

    return proto_content, proto_filename
